Implement adding/removing policies to application, orderer, and consortium organizations

stephyee opened the pull request:
#### Type of change

- New feature

#### Description

Support adding and removing policies from either an application, orderer, or consortium organization.

#### Related issues

[FAB-17570](https://jira.hyperledger.org/browse/FAB-17570)
[FAB-17571](https://jira.hyperledger.org/browse/FAB-17571)
[FAB-17572](https://jira.hyperledger.org/browse/FAB-17572)

stephyee commented:
/azp run

stephyee commented:
/ci-run

caod123 commented:
For simplicity, do you think you could move all the policy related API/functionality to a separate policy.go?

stephyee committed:
[FAB-17570] Implement adding/removing policies from application

- Adding a policy to application or application org will not error if policy exists
- Removing a policy from an application or application org will error if policy
does not exist

Signed-off-by: Tiffany Harris <tiffany.harris@ibm.com>

stephyee committed:
[FAB-17572] Implement adding/removing policies from consortiums orgs

Signed-off-by: Tiffany Harris <tiffany.harris@ibm.com>

stephyee committed:
[FAB-17571] Implement adding/removing policies from orderer

- Adding a policy to orderer or orderer org will not error if policy exists
- Removing a policy from an orderer or orderer org will error if policy
    does not exist
- Removing the BlockValidation policy is not allowed

Signed-off-by: Tiffany Harris <tiffany.harris@ibm.com>

DereckLuo reviewed (commented):
LGTM

stephyee commented:
This may have introduced a race - investigating.

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -287,3 +287,154 @@ func TestRemoveApplicationPolicyFailures(t *testing.T) {
+
+	err = AddConsortiumOrgPolicy(config, "Consortium1", "Org1", "TestPolicy", Policy{Type: ImplicitMetaPolicyType, Rule: "MAJORITY Endorsement"})
+	gt.Expect(err).NotTo(HaveOccurred())
+	gt.Expect(len(config.ChannelGroup.Groups[ConsortiumsGroupKey].Groups["Consortium1"].Groups["Org1"].Policies)).To(Equal(5))
```
We have a matcher for this: `Expect(slice).To(HaveLen(5))`

Please add a follow-up task for that.

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -287,3 +287,154 @@ func TestRemoveApplicationPolicyFailures(t *testing.T) {
+	err = AddConsortiumOrgPolicy(config, "Consortium1", "Org1", "TestPolicy", Policy{Type: ImplicitMetaPolicyType, Rule: "MAJORITY Endorsement"})
+	gt.Expect(err).NotTo(HaveOccurred())
+	gt.Expect(len(config.ChannelGroup.Groups[ConsortiumsGroupKey].Groups["Consortium1"].Groups["Org1"].Policies)).To(Equal(5))
+	gt.Expect(config.ChannelGroup.Groups[ConsortiumsGroupKey].Groups["Consortium1"].Groups["Org1"].Policies[AdminsPolicyKey]).NotTo(BeNil())
```
This is an extremely long chain to assert on.

`config.ChannelGroup.Groups[ConsortiumsGroupKey].Groups["Consortium1"].Groups["Org1"].Policies[AdminsPolicyKey`

Given you shadowed a (relatively) short constant in the production code, you should consider using a variable for this in the assertions.

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -287,3 +287,154 @@ func TestRemoveApplicationPolicyFailures(t *testing.T) {
+
+	err = RemoveConsortiumOrgPolicy(config, "Consortium1", "Org1", "TestPolicy")
+	gt.Expect(err).NotTo(HaveOccurred())
+	gt.Expect(len(config.ChannelGroup.Groups[ConsortiumsGroupKey].Groups["Consortium1"].Groups["Org1"].Policies)).To(Equal(4))
+	gt.Expect(config.ChannelGroup.Groups[ConsortiumsGroupKey].Groups["Consortium1"].Groups["Org1"].Policies[AdminsPolicyKey]).NotTo(BeNil())
```
Same comments as the other test; HaveLen matcher and a variable to avoid re-typing the reference chain _n_ times.

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -438,3 +438,252 @@ func TestRemoveConsortiumOrgPolicyFailures(t *testing.T) {
+
+	err = AddOrdererPolicy(config, AdminsPolicyKey, "TestPolicy", Policy{Type: ImplicitMetaPolicyType, Rule: "ANY Endorsement"})
+	gt.Expect(err).NotTo(HaveOccurred())
+	gt.Expect(len(config.ChannelGroup.Groups[OrdererGroupKey].Policies)).To(Equal(5))
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Policies[AdminsPolicyKey]).NotTo(BeNil())
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Policies[ReadersPolicyKey]).NotTo(BeNil())
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Policies[WritersPolicyKey]).NotTo(BeNil())
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Policies["TestPolicy"]).NotTo(BeNil())
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Policies[BlockValidationPolicyKey]).NotTo(BeNil())
+}
```
Same comments as earlier. (Applies the first part of the path up to `Policies` this time.)

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -438,3 +438,252 @@ func TestRemoveConsortiumOrgPolicyFailures(t *testing.T) {
+
+	err = RemoveOrdererPolicy(config, "TestPolicy")
+	gt.Expect(err).NotTo(HaveOccurred())
+	gt.Expect(len(config.ChannelGroup.Groups[OrdererGroupKey].Groups["OrdererOrg"].Policies)).To(Equal(4))
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Groups["OrdererOrg"].Policies[AdminsPolicyKey]).NotTo(BeNil())
```
ditto

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -438,3 +438,252 @@ func TestRemoveConsortiumOrgPolicyFailures(t *testing.T) {
+
+	err = AddOrdererOrgPolicy(config, "OrdererOrg", AdminsPolicyKey, "TestPolicy", Policy{Type: ImplicitMetaPolicyType, Rule: "ANY Endorsement"})
+	gt.Expect(err).NotTo(HaveOccurred())
+	gt.Expect(len(config.ChannelGroup.Groups[OrdererGroupKey].Groups["OrdererOrg"].Policies)).To(Equal(5))
+	gt.Expect(config.ChannelGroup.Groups[OrdererGroupKey].Groups["OrdererOrg"].Policies[AdminsPolicyKey]).NotTo(BeNil())
```
Ditto

sykesm commented on pkg/config/policies_test.go:
```diff
@@ -0,0 +1,289 @@
+	gt.Expect(err).NotTo(HaveOccurred())
+
+	actualPolicies := config.ChannelGroup.Groups[ApplicationGroupKey].Policies
+	gt.Expect(len(actualPolicies)).To(Equal(3))
```
Please use the HaveLen matcher.

sykesm reviewed (approved):
LGTM. Minor nits for folllow-up related to repeated use of long-reference chains in tests and usage of `len(slice)..Equal` vs. `HaveLen` matcher.
